File: scripts/test_solutions/416_partition_equal_subset_sum.py
def canPartition(nums):
    total = sum(nums)
    if total % 2 != 0:
        return False
    target = total // 2
    dp = {0}
    for num in nums:
        new_dp = set()
        for val in dp:
            new_dp.add(val)
            new_dp.add(val + num)
        if target in new_dp:
            return True
        dp = new_dp
    return target in dp

File: scripts/test_solutions/416_partition_equal_subset_sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int total = 0;
        for (int num : nums) total += num;
        if (total % 2 != 0) return false;

        int target = total / 2;
        unordered_set<int> dp;
        dp.insert(0);

        for (int num : nums) {
            unordered_set<int> newDp;
            for (int val : dp) {
                newDp.insert(val);
                newDp.insert(val + num);
            }
            if (newDp.count(target)) return true;
            dp = newDp;
        }
        return dp.count(target);
    }
};
